File: hannah_bot/hannah_bot/simple_parameter.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from rcl_interfaces.msg import SetParametersResult
from rclpy.parameter import Parameter

class SimpleParameter(Node):
    def __init__(self):
        super().__init__("SimplePublisher")
        self.declare_parameter("simple_int_param", 28)
        self.declare_parameter("simple_string_param", "Ferbin")
        self.add_on_set_parameters_callback(self.paramchange_callback)

    def paramchange_callback(self, params):
        result = SetParametersResult(successful=True)  # Initialize result as successful.

        for param in params:
            if param.name == "simple_int_param" and param.type_ == Parameter.Type.INTEGER:
                self.get_logger().info("Simple int parameter changed! with name set to %d" % param.value)
                # Use param.value to access the parameter value.
                result.successful = True

            if param.name == "simple_string_param" and param.type_ == Parameter.Type.STRING:
                self.get_logger().info("Simple string parameter changed! with name %s set to %s" % (param.name, param.value))
                # Use param.name and param.value to access the parameter name and value.
                result.successful = True

        return result

def main():
    rclpy.init()  # Instantiate the communication with ROS
    simple_parameter = SimpleParameter()
    rclpy.spin(simple_parameter)
    simple_parameter.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: hannah_cpp_egs/src/simple_parameter.cpp
#include <rclcpp/rclcpp.hpp>
#include <rcl_interfaces/msg/set_parameters_result.hpp>
#include <string>
#include <vector>

using std::placeholders::_1;

class SimpleParameter : public rclcpp::Node
{
public:
    SimpleParameter() : Node("simple_parameter")
    {
        declare_parameter<int>("simple_init_param", 28);
        declare_parameter<std::string>("simple_string_param", "FERBIN");

        // Use the correct function name: add_on_set_parameters_callback
        param_callback_handle_ = add_on_set_parameters_callback(std::bind(&SimpleParameter::paramChangeCallback, this, _1));
    }

private:
    rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;

    rcl_interfaces::msg::SetParametersResult paramChangeCallback(const std::vector<rclcpp::Parameter> &parameters)
    {
        rcl_interfaces::msg::SetParametersResult result;
        for (const auto &param : parameters)
        {
            if (param.get_name() == "simple_init_param" && param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER)
            {
                RCLCPP_INFO(get_logger(), "Param simple_init_param changed: %ld", param.as_int());
                result.successful = true;
            }

            if (param.get_name() == "simple_string_param" && param.get_type() == rclcpp::ParameterType::PARAMETER_STRING)
            {
                RCLCPP_INFO(get_logger(), "Param simple_string_param changed: %s", param.as_string().c_str());
                result.successful = true;
            }
        }
        return result;
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleParameter>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
